src/path.cr

```diff
@@ -0,0 +1,941 @@
+          end
+        end
+      end
+
+      if str.empty?
+        str << '.'
+      end
+
+      last_char = (str.buffer + str.bytesize - 1).value.unsafe_chr
+
+      if add_separator_at_end && !separators.includes?(last_char)
+        str << separators[0]
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Yields each component of this path as a `String`.
+  def each_part
+    last_pos = 0
+    each_part_separator_index do |pos|
+      yield @name.byte_slice(last_pos, pos - last_pos)
+      last_pos = pos
+    end
+  end
+
+  # Returns the components of this path as an `Array(String)`.
+  def parts : Array(String)
+    parts = [] of String
+    each_part do |part|
+      parts << part
+    end
+    parts
+  end
+
+  private def each_part_separator_index
+    reader = Char::Reader.new(@name)
+    last_was_separator = false
+    reader.each do |char|
+      if separators.includes?(char)
+        yield reader.pos unless last_was_separator
+        last_was_separator = true
+      else
+        last_was_separator = false
+      end
+    end
+  end
+
+  # Converts this path to a Windows path.
+  #
+  # ```
+  # Path.posix("foo/bar").to_windows   # => Path.windows("foo/bar")
+  # Path.windows("foo/bar").to_windows # => Path.windows("foo/bar")
+  # ```
+  #
+  # This creates a new instance with the same string representation but with
+  # `Kind::WINDOWS`.
+  def to_windows : Path
+    new_instance(@name, Kind::WINDOWS)
+  end
+
+  # Converts this path to a POSIX path.
+  #
+  # ```
+  # Path.windows("foo/bar\\baz").to_posix # => Path.posix("foo/bar/baz")
+  # Path.posix("foo/bar").to_posix        # => Path.posix("foo/bar")
+  # Path.posix("foo/bar\\baz").to_posix   # => Path.posix("foo/bar\\baz")
+  # ```
+  #
+  # It returns a copy of this instance if it already has POSIX kind. Otherwise
+  # a new instance is created with `Kind::POSIX` and all occurences of
+  # backslash file separators (`\\`) replaced by forward slash (`/`).
+  def to_posix : Path
+    if posix?
+      new_instance(@name, Kind::POSIX)
+    else
+      new_instance(@name.gsub(Path.separators(Kind::WINDOWS)[0], Path.separators(Kind::POSIX)[0]), Kind::POSIX)
+    end
+  end
+
+  # Converts this path to the fiven *kind*.
+  #
+  # See `#to_windows` and `#to_posix` for details.
+  def to_kind(kind)
+    if Kind.resolve_native(kind).posix?
+      to_posix
+    else
+      to_windows
+    end
+  end
+
+  # Converts this path to an absolute path. Relative paths are
+  # referenced from the current working directory of the process (`Dir.current`)
+  # unless *base* is given, in which case it will be used as the reference path.
+  #
+  # ```
+  # Path{"foo"}.expand             # => Path{"/current/path/foo"}
+  # Path{"~/crystal/foo"}.expand   # => Path{"/home/crystal/foo"}
+  # Path{"baz"}.expand("/foo/bar") # => Path{"/foo/bar/baz"}
+  # ```
+  def expand(base : Path | String = Dir.current, home = Path.home, expand_base = true) : Path
```

Comment:
`expand_base` and `home` are not described in the docs.
Maybe `*` should be used for force named arguments in this api.